stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt

```diff
@@ -54,4 +59,126 @@ public object StreamFileUtil {
             null
         }
     }
+
+    /**
+     * Creates a Stream cache directory if one doesn't exist already.
+     *
+     * @param context The [Context] necessary to perform
+     * file operations.
+     * @param onError Called when an error occurs.
+     *
+     * @return The Stream cache directory if it exists or its creation
+     * was successful, false otherwise.
+     */
+    @Suppress("TooGenericExceptionCaught")
+    private fun getOrCreateStreamCacheDir(
+        context: Context,
+        onError: (exception: Exception) -> Unit = {},
+    ): File? {
+        return try {
+            File(context.cacheDir, streamCacheDirName).also { streamCacheDir ->
+                streamCacheDir.mkdirs()
+            }
+        } catch (e: Exception) {
+            onError(e)
+            null
+        }
+    }
+
+    /**
+     * Deletes all the content contained within the
+     * Stream cache directory.
+     *
+     * @param context The [Context] necessary to perform
+     * file operations.
+     * @param onError Called when an error occurs.
+     *
+     * @return True if the cache was cleared, false otherwise.
+     */
+    @Suppress("TooGenericExceptionCaught")
+    public fun clearStreamCache(
+        context: Context,
+        onError: (exception: Exception) -> Unit = {},
+    ): Boolean {
+        return try {
+            val directory = File(context.cacheDir, streamCacheDirName)
+
+            directory.listFiles()?.forEach {
+                it.delete()
+            }
+
+            true
+        } catch (e: Exception) {
+            onError(e)
+            false
+        }
+    }
+
+    /**
+     * Hashes the links of given attachments and then tries to create a new file
+     * under that hash. If the file already exists checks that the full file
+     * has been written and shares it if it has, in other cases downloads the file
+     * and writes it.
+     *
+     * @param context The Android [Context] used for path resolving and [Uri] fetching.
+     * @param attachment the attachment to be downloaded.
+     * @param onError Called when an error occurs.
+     *
+     * @return The [Uri] that represents the path to the downloaded file.
+     */
+    public suspend fun writeFileToShareableFile(
+        context: Context,
+        attachment: Attachment,
+        onError: (exception: Exception) -> Unit = {},
+    ): Uri? {
+        val result = runCatching {
+            val streamCacheDir = getOrCreateStreamCacheDir(context, onError)
+            val attachmentName = (attachment.url ?: attachment.assetUrl)?.hashCode()
+            val fileName = attachmentName.toString() + attachment.name
+
+            val file = File(streamCacheDir, fileName)
```

Comment:
If `getOrCreateStreamCacheDir()` fails, we will be notified twice on `onError` callback, one because it couldn't be created (But the execution wasn't stopped) and another one when the File will be trying to be created (There is no parent folder....)
I don't see any advantage of having `onError` functions (Even when it is empty implementation)
We should:
1 -. Throw exception (We should be clear about which exceptions can be throw here)
2-. Return an Either type with the `Uri` value or `Exception`

Reply:
Removed the `onError` functions and adopted using `Result<T>` in order to make exception handling clearer, as discussed in the meeting.